File: backend/app/models/download.py
from typing import Optional
from pydantic import BaseModel, Field


class DownloadRequest(BaseModel):
    url: str
    type: str = Field(default="video", description="video or audio")
    quality: str = Field(default="best", description="e.g. best, 2160, 1440, 1080, 720, 480, 360 or 320, 256, 192, 128")
    format: str = Field(default="mp4", description="mp4, webm, mkv, mp3, m4a, opus, wav")
    save_path: Optional[str] = None
    filename_template: Optional[str] = "%(title)s.%(ext)s"
    subtitle_lang: Optional[str] = None # None, en, th, orig, all
    embed_subtitle: bool = False
    embed_thumbnail: bool = False
    embed_metadata: bool = True


class ProgressEvent(BaseModel):
    download_id: str
    status: str # queued, analyzing, downloading, processing, merging, converting, completed, failed, cancelled
    progress: float = 0.0 # 0.0 to 100.0
    speed: Optional[float] = None # bytes per second
    speed_string: Optional[str] = None # e.g. "12.5 MB/s"
    eta: Optional[int] = None # seconds
    eta_string: Optional[str] = None # e.g. "00:04"
    downloaded_bytes: Optional[int] = None
    total_bytes: Optional[int] = None
    file_path: Optional[str] = None
    filename: Optional[str] = None
    error_message: Optional[str] = None


class DownloadRecord(BaseModel):
    id: str
    video_id: str
    url: str
    title: str
    channel: str
    thumbnail: str
    format: str
    quality: str
    file_path: str
    file_size: Optional[int] = None
    status: str
    created_at: str
    completed_at: Optional[str] = None
    error_message: Optional[str] = None

File: frontend/types/download.ts
export interface DownloadRequest {
  url: string;
  type: "video" | "audio";
  quality: string;
  format: string;
  save_path?: string;
  filename_template?: string;
  subtitle_lang?: string;
  embed_subtitle?: boolean;
  embed_thumbnail?: boolean;
  embed_metadata?: boolean;
}

export interface ProgressEvent {
  download_id: string;
  status: string; // queued, analyzing, downloading, processing, merging, converting, completed, failed, cancelled
  progress: number; // 0.0 to 100.0
  speed?: number;
  speed_string?: string;
  eta?: number;
  eta_string?: string;
  downloaded_bytes?: number;
  total_bytes?: number;
  file_path?: string;
  filename?: string;
  error_message?: string;
}

export interface AppSettings {
  default_folder: string;
  default_video_quality: string;
  default_video_format: string;
  default_audio_quality: string;
  default_audio_format: string;
  concurrent_downloads: number;
  embed_metadata: boolean;
  embed_thumbnail: boolean;
  clipboard_detection: boolean;
  filename_template: string;
  ffmpeg_path: string;
  ytdlp_path: string;
  theme: string;
}

export interface SystemStatus {
  ytdlp_installed: boolean;
  ytdlp_version: string;
  ffmpeg_installed: boolean;
  ffmpeg_path: string;
  ffmpeg_version: string;
  download_folder: string;
  download_folder_writable: boolean;
  disk_free_bytes: number;
  disk_free_gb: number;
  os_name: string;
}

export interface CommonFolder {
  name: string;
  path: string;
}
